matrix-client.el

```diff
@@ -180,113 +199,104 @@ event-handlers and input-filters.")
   (when (or force matrix-client-enable-watchdog)
-    (let ((last-ts (or (and (slot-boundp con :last-event-ts)
-                            (oref con :last-event-ts))
-                       0))
-          (next (and (slot-boundp con :end-token)
-                     (oref con :end-token)))
-          (timer (and (slot-boundp con :watchdog-timer)
-                      (oref con :watchdog-timer))))
-      (if (and (slot-boundp con :watchdog-timer) ;; start timer if not running
-               (oref con :watchdog-timer))
-          (if (> (* 1000 (- (float-time) last-ts)) ;; If we've timed out, re-sync
+    (let ((last-ts (oref con :last-event-ts))
+          (next (oref con :end-token))
+          (timer (oref con :watchdog-timer)))
+      (if timer
+          (if (> (* 1000 (- (float-time) last-ts))
                  matrix-client-event-poll-timeout)
               (progn
+                ;; Timed out: resync
                 (cancel-timer timer)
                 ;; XXX Pull these fucking syncs out and bar them on (oref con :running)
-                (when (and (slot-boundp con :running)
-                           (oref con :running))
-                  (message "Reconnecting you to Matrix, one monent please.")
+                (when (oref con :running)
+                  (message "Reconnecting you to Matrix, one moment please...")
                   (cancel-timer timer)
                   (matrix-sync con next nil matrix-client-event-poll-timeout
                                (apply-partially #'matrix-client-sync-handler con))))
+            ;; Not timed out: just cancel timer
             (cancel-timer timer)))
-      (oset con :watchdog-timer
-            (run-with-timer (or timer-secs (/ (* 2 matrix-client-event-poll-timeout) 1000))
-                            (or timer-secs (/ (* 2 matrix-client-event-poll-timeout) 1000))
-                            (apply-partially #'matrix-client-start-watchdog con))))))
+      (let ((timer-secs (or timer-secs (/ (* 2 matrix-client-event-poll-timeout) 1000))))
+        (oset con :watchdog-timer (run-with-timer timer-secs timer-secs
+                                                  (apply-partially #'matrix-client-start-watchdog con)))))))
 
-(defmethod matrix-client-setup-room ((con matrix-client-connection) room-id)
+(cl-defmethod matrix-client-setup-room ((con matrix-client-connection) room-id)
+  "Prepare and switch to buffer for ROOM-ID, and return room object."
   (when (get-buffer room-id)
     (kill-buffer room-id))
   (let* ((room-buf (get-buffer-create room-id))
-         (room-obj (matrix-client-room room-id :buffer room-buf :con con))
-         (new-room-list (append (oref con :rooms) (list (cons room-id room-obj)))))
+         (room-obj (matrix-client-room room-id :buffer room-buf :con con)))
     (with-current-buffer room-buf
       (matrix-client-mode)
+      (setq buffer-undo-list t)
```

Comment:
Was undo not working fine in buffers? I wouldn't be surprised, so I guess it is best to disable it completely.

Reply:
I don't think we want undo enabled in chat room buffers, because we wouldn't want to erase chat messages if the user hit undo.  The code already disabled undo, but it did it in a less optimal way, by clearing the undo buffer after every insert.  By setting it to `t` when the buffer is created, it's disabled from the beginning.

Reply:
ah, I see, this makes perfect sense now :)